libpod/container_internal.go

```diff
@@ -1103,3 +1108,58 @@ func (c *Container) saveSpec(spec *spec.Spec) error {
 
 	return nil
 }
+
+func (c *Container) setupOCIHooks(g *generate.Generator) error {
+	addedHooks := map[string]struct{}{}
+	ocihooks, err := hooks.SetupHooks(c.runtime.config.HooksDir)
+	if err != nil {
+		return err
+	}
+	addHook := func(hook hooks.HookParams) error {
+		// Only add a hook once
+		if _, ok := addedHooks[hook.Hook]; !ok {
+			if err := hooks.AddOCIHook(g, hook); err != nil {
+				return err
+			}
+			addedHooks[hook.Hook] = struct{}{}
+		}
+		return nil
+	}
+	for _, hook := range ocihooks {
+		logrus.Debugf("SetupOCIHooks", hook)
+		if hook.HasBindMounts && len(c.config.Spec.Mounts) > 0 {
```

Comment:
I'm looking at porting kubernetes-incubator/cri-o#1517 to libpod now that kubernetes-incubator/cri-o#1537 has demonstrated vendoring libpod code there, and this line doesn't look right to me.  Shouldn't we be walking `Mounts` looking for `bind` or `rbind` in `Options`?  As it stands this is more like `hasMountsAtAll` instead of `hasBindMounts`.  Can I fix that in my hook reroll, or do we need to preserve the has-mounts-at-all logic here for some reason?

Reply:
No I think we should be looking at bind mounts, this is a bug.

Reply:
@wking `c.state.BindMounts` represents mounts that we add to the config before writing it to disk. They are not present in `c.config.Spec.Mounts` which holds the spec the user passed in.

`libpod` accepts an OCI spec when creating a container and uses it mostly unmodified, but we add a few environment variables, annotations, mounts (the aforementioned `/dev/shm` among others), as well as setting the location of the rootfs (not known before we create the container and mount the image)

Reply:
@wking `libpod` doesn't know about `--volume` - it can only distinguish between mounts in the passed-in OCI spec, and mounts it adds itself. Mounts added within `libpod` are stored in `c.state.BindMounts` while the OCI spec, `c.config.Spec`, holds the mounts passed in with the spec. The spec itself is created by parsing the `podman create` or `podman run` command line, and I don't believe it adds any bind mounts not requested by the user via `--volume` (though it's been awhile since I looked through that code).

Reply:
For image volumes - those presently are added separately from both `c.state.BindMounts` and the OCI spec, though I want to refactor them to be stored in `c.state.BindMounts` as well. They are added at the same time as `c.state.BindMounts` is added to the spec, but they are put directly into the final spec we write to disk, skipping `c.state.BindMounts`

Reply:
That perhaps came out a bit more confusing than intended, so let me clarify...

We accept an OCI spec when we create a container in `libpod`. This is not the final spec we use to create the container, as `libpod` adds a few things to it before writing it to disk (`generateSpec()` produces the final spec we write). We add a few mounts during this stage - those included in `c.state.BindMounts` and image volume mounts (if requested). These are never present in `c.config.Spec` which is immutable after being passed in by the user, but will be in the spec we write to disk.

Reply:
Well the object is only to cover --volume mounts (bind) mounts. oci-umount is looking at anything handed into the container runtime that is mounted in addition to the mount points that the runtime creates.

Reply:
Fix in flight with #700.